File: app/src/main/java/com/qadam/lastpuff/domain/support/SosMessageComposer.kt
package com.qadam.lastpuff.domain.support

object SosMessageComposer {

    private val milestones = listOf(
        0 to MessageCategory.SOS_START,
        15 to MessageCategory.SOS_15_SEC,
        30 to MessageCategory.SOS_30_SEC,
        45 to MessageCategory.SOS_45_SEC,
        60 to MessageCategory.SOS_60_SEC,
        75 to MessageCategory.SOS_75_SEC,
        90 to MessageCategory.SOS_90_SEC,
        105 to MessageCategory.SOS_105_SEC,
        120 to MessageCategory.SOS_120_SEC,
        135 to MessageCategory.SOS_140_SEC,
        150 to MessageCategory.SOS_LAST_20,
        165 to MessageCategory.SOS_LAST_20,
        180 to MessageCategory.SOS_LAST_20
    )

    fun compose(
        elapsedSeconds: Int,
        ctx: SupportContext,
        session: MessageSession,
        mode: SosMode,
        flags: SosComposeFlags
    ): SosMoment {
        val category = milestones.lastOrNull { (sec, _) -> elapsedSeconds >= sec }?.second
            ?: MessageCategory.SOS_START

        val primary = session.pick(category)
        var secondary: String? = null
        val action = session.pickAction(mode)
        var showLetter = false

        when (elapsedSeconds) {
            30 -> if (!flags.factShown) {
                secondary = session.pick(MessageCategory.FACT)
            }
            45 -> secondary = session.pick(MessageCategory.FUTURE_HOLD)
            60 -> {
                if (ctx.pricePerCigarette > 0) {
                    val saved = ctx.pricePerCigarette.coerceAtLeast(1.0)
                    secondary = "Пока ты держишься, ты уже сохранил ещё ${saved.toInt()} ${ctx.currency}."
                } else {
                    secondary = session.pick(MessageCategory.MONEY)
                }
            }
            75 -> {
                val reason = ctx.reasons.firstOrNull()
                val memory = reason?.let { SupportMessageBank.memoryForReason(it, ctx.goalTitle) }
                if (!ctx.personalLetter.isNullOrBlank() && !flags.letterShown) {
                    showLetter = true
                } else if (memory != null) {
                    secondary = memory
                } else {
                    secondary = session.pick(MessageCategory.FAMILY)
                }
            }
            90 -> {
                secondary = if (ctx.totalWins >= 3) {
                    "Ты уже победил ${ctx.totalWins} приступов.\nПочему этот должен быть другим?"
                } else {
                    session.pick(MessageCategory.PSYCHOLOGY)
                }
            }
            105 -> secondary = session.pick(MessageCategory.PSYCHOLOGY)
            120 -> secondary = session.pickHumorRarely() ?: session.pick(MessageCategory.DIALOG)
            135 -> secondary = session.pick(MessageCategory.ACHIEVEMENT_MOMENT)
        }

        return SosMoment(
            primary = primary,
            secondary = secondary,
            action = action,
            showPersonalLetter = showLetter
        )
    }

    fun victoryMessage(session: MessageSession): String = session.pick(MessageCategory.VICTORY)

    fun relapseMessage(session: MessageSession): String = session.pick(MessageCategory.RELAPSE)
}

data class SosComposeFlags(
    val factShown: Boolean = false,
    val letterShown: Boolean = false
)

File: ios/QadamLastPuff/Domain/SosMessageComposer.swift
import Foundation

enum SosMessageComposer {
    private static let milestones: [(Int, MessageCategory)] = [
        (0, .sosStart), (15, .sos15Sec), (30, .sos30Sec), (45, .sos45Sec),
        (60, .sos60Sec), (75, .sos75Sec), (90, .sos90Sec), (105, .sos105Sec),
        (120, .sos120Sec), (135, .sos140Sec), (150, .sosLast20),
        (165, .sosLast20), (180, .sosLast20)
    ]

    static func compose(
        elapsedSeconds: Int,
        ctx: SupportContext,
        session: MessageSession,
        mode: SosMode,
        flags: SosComposeFlags
    ) -> SosMoment {
        let category = milestones.last { elapsedSeconds >= $0.0 }?.1 ?? .sosStart
        let primary = session.pick(category)
        var secondary: String?
        let action = session.pickAction(mode: mode)
        var showLetter = false

        switch elapsedSeconds {
        case 30:
            if !flags.factShown { secondary = session.pick(.fact) }
        case 45:
            secondary = session.pick(.futureHold)
        case 60:
            if ctx.moneySaved >= 1 {
                secondary = "Пока ты держишься, ты уже сэкономил \(Int(ctx.moneySaved)) \(ctx.currency)."
            } else if ctx.pricePerCigarette > 0 {
                secondary = session.pick(.money)
            } else {
                secondary = session.pick(.money)
            }
        case 75:
            let reason = ctx.reasons.first
            let memory = reason.flatMap { SupportMessageBank.memoryForReason($0, goalTitle: ctx.goalTitle) }
            if let letter = ctx.personalLetter, !letter.isEmpty, !flags.letterShown {
                showLetter = true
            } else if let memory {
                secondary = memory
            } else {
                secondary = session.pick(.family)
            }
        case 90:
            secondary = ctx.totalWins >= 3
                ? "Ты уже победил \(ctx.totalWins) приступов.\nПочему этот должен быть другим?"
                : session.pick(.psychology)
        case 105:
            secondary = session.pick(.psychology)
        case 120:
            secondary = session.pickHumorRarely() ?? session.pick(.dialog)
        case 135:
            secondary = session.pick(.achievementMoment)
        default:
            break
        }

        return SosMoment(primary: primary, secondary: secondary, action: action, showPersonalLetter: showLetter)
    }

    static func victoryMessage(session: MessageSession) -> String {
        session.pick(.victory)
    }

    static func relapseMessage(session: MessageSession) -> String {
        session.pick(.relapse)
    }
}
